validator/sawtooth_validator/consensus/registry.py

```diff
@@ -45,27 +46,46 @@ def has_active_engine(self):
         with self._lock:
             return self._active is not None
 
-    def register_engine(self, connection_id, name, version):
+    def register_engine(self, connection_id, name, version,
+                        additional_protocols):
         with self._lock:
-            # If this engine is already registered, remove the old connection
+            engine_info = EngineInfo(
+                connection_id, name, version, additional_protocols)
+
+            # If this engine is already registered or if there is an existing
+            # engine that this one replaces (supports all the same protocols),
+            # remove the old connection.
+            def is_same_engine(eng):
+                return eng.name == name and eng.version == version
+
+            def is_replaced(eng):
+                return (
+                    engine_handles_protocol(eng.name, eng.version, engine_info)
+                    and all([True for (n, v) in eng.additional_protocols
+                             if engine_handles_protocol(n, v, engine_info)]))
+
             self._registry = list(filter(
-                lambda e: e.name != name and e.version != version,
+                lambda e: not is_same_engine(e) and not is_replaced(e),
```

Comment:
Can you add some tests for these scenarios?

Reply:
Added registry tests in a new commit